File: kinv/core/consolidate/findings.py
"""The consolidation report: duplicates, near-values, and the question that started it all."""

from __future__ import annotations

from typing import Iterable

from kinv.core.canonical import search_label
from kinv.core.units import format_magnitude
from kinv.js import code_units, collation_key, js_sorted, locale_compare

__all__ = [
    "FIELD_JOIN",
    "compare_columns",
    "consolidate",
    "distinct_parts",
    "parts_by_key",
]


def _placements(lines: Iterable[dict]) -> int:
    return sum(len(line["line"]["refs"]) for line in lines)


def _refs(lines: Iterable[dict]) -> list[str]:
    return [ref for line in lines for ref in line["line"]["refs"]]


def _group_by_key(lines: Iterable[dict]) -> list[dict]:
    buckets: dict[str, list[dict]] = {}
    for line in lines:
        buckets.setdefault(line["key"], []).append(line)
    return [
        {"key": key, "lines": ls, "placements": _placements(ls), "refs": _refs(ls)}
        for key, ls in buckets.items()
    ]


def _value_key(spec: dict) -> str:
    """Generic specs only, keyed by class and magnitude so packages can be compared."""
    return f"{spec['cls']}|{format_magnitude(spec['magnitude'])}"


def _by(lines: Iterable[dict], field) -> dict[str, list[dict]]:
    buckets: dict[str, list[dict]] = {}
    for line in lines:
        buckets.setdefault(field(line), []).append(line)
    return buckets


def _duplicate_spellings(groups: list[dict]) -> list[dict]:
    out = []
    for group in groups:
        by_value = _by(group["lines"], lambda line: line["line"]["value"].strip())
        if len(by_value) < 2:
            continue
        spellings = [
            {"value": value, "placements": _placements(ls), "refs": _refs(ls)}
            for value, ls in by_value.items()
        ]
        out.append(
            {
                "kind": "duplicate-spelling",
                "key": group["key"],
                "spellings": js_sorted(spellings, lambda a, b: b["placements"] - a["placements"]),
                "placements": group["placements"],
                "linesSaved": len(by_value) - 1,
            }
        )
    return out


def _mixed_footprints(groups: list[dict]) -> list[dict]:
    out = []
    for group in groups:
        by_footprint = _by(group["lines"], lambda line: line["line"]["footprint"].strip())
        if len(by_footprint) < 2:
            continue
        footprints = [
            {"footprint": footprint, "placements": _placements(ls), "refs": _refs(ls)}
            for footprint, ls in by_footprint.items()
        ]
        out.append(
            {
                "kind": "mixed-footprint",
                "key": group["key"],
                "footprints": js_sorted(footprints, lambda a, b: b["placements"] - a["placements"]),
            }
        )
    return out


def _multi_package(lines: list[dict]) -> list[dict]:
    by_value = _by((line for line in lines if line["spec"]["kind"] == "generic"), lambda l: _value_key(l["spec"]))

    out = []
    for ls in by_value.values():
        by_pkg = _by(ls, lambda line: line["spec"]["pkg"] if line["spec"]["pkg"] is not None else "?")
        if len(by_pkg) < 2:
            continue

        groups = js_sorted(
            [{"pkg": pkg, "placements": _placements(group), "refs": _refs(group)} for pkg, group in by_pkg.items()],
            lambda a, b: (b["placements"] - a["placements"]) or locale_compare(a["pkg"], b["pkg"]),
        )
        spec = ls[0]["spec"]
        out.append(
            {
                "kind": "multi-package",
                "cls": spec["cls"],
                "value": format_magnitude(spec["magnitude"]),
                "unit": spec["unit"],
                "groups": groups,
                "suggested": groups[0]["pkg"],
                # A tie means the tool has no basis to prefer either size; say so
                # rather than presenting an arbitrary pick as a recommendation.
                "tied": len(groups) > 1 and groups[0]["placements"] == groups[1]["placements"],
                "linesSaved": len(groups) - 1,
            }
        )
    return out


def _magnitudes_by_class(lines: list[dict]) -> dict[str, dict[float, dict]]:
    """By class and then by *magnitude*, deliberately not by key.

    10k in 0402 and 10k in 0603 are the multi-package finding, not a
    near-value one.
    """
    by_class: dict[str, dict[float, dict]] = {}
    for line in lines:
        spec = line["spec"]
        # 0R has no meaningful neighbourhood: everything is infinitely far from it.
        if spec["kind"] != "generic" or spec["magnitude"] == 0:
            continue
        magnitudes = by_class.setdefault(spec["cls"], {})
        slot = magnitudes.setdefault(spec["magnitude"], {"refs": [], "keys": {}})
        slot["refs"].extend(line["line"]["refs"])
        slot["keys"][line["key"]] = True  # an insertion-ordered set
    return by_class


def _near_values(lines: list[dict], percent: float) -> list[dict]:
    out = []
    for cls, magnitudes in _magnitudes_by_class(lines).items():
        ordered = js_sorted(magnitudes.items(), lambda a, b: a[0] - b[0])

        # Walk the sorted values, collecting runs where each step is within
        # tolerance of the last. 49.9k -> 50k -> 51k chains into one finding.
        run: list[tuple[float, dict]] = []

        def flush() -> None:
            if len(run) >= 2:
                lo, hi = run[0][0], run[-1][0]
                out.append(
                    {
                        "kind": "near-value",
                        "cls": cls,
                        "members": [
                            {
                                "key": " ".join(sorted(slot["keys"], key=code_units)),
                                "value": format_magnitude(magnitude),
                                "placements": len(slot["refs"]),
                                "refs": slot["refs"],
                            }
                            for magnitude, slot in run
                        ],
                        "spreadPercent": ((hi - lo) / lo) * 100,
                    }
                )

        for entry in ordered:
            if run and ((entry[0] - run[-1][0]) / run[-1][0]) * 100 <= percent:
                run.append(entry)
            else:
                flush()
                run = [entry]
        flush()
    return out


def _singletons(groups: list[dict]) -> list[dict]:
    out = []
    for group in groups:
        if group["placements"] != 1:
            continue
        line = group["lines"][0]
        # Only generics: you can swap a lone 82k for an 80.6k you already
        # stock, but a board with one RP2040 is not a consolidation opportunity.
        if line["spec"]["kind"] != "generic":
            continue
        out.append(
            {
                "kind": "singleton",
                "key": group["key"],
                "cls": line["spec"]["cls"],
                "value": line["line"]["value"].strip(),
                "ref": group["refs"][0],
            }
        )
    return out


def consolidate(lines: Iterable[dict], near_value_percent: float | None = None) -> list[dict]:
    percent = 2 if near_value_percent is None else near_value_percent
    # Excluded lines never reach the report: a board with 8 test points should
    # not produce eight "used once" findings about pads you cannot buy.
    buyable = [line for line in lines if line["excluded"] is None]
    groups = _group_by_key(buyable)
    return [
        *_duplicate_spellings(groups),
        *_mixed_footprints(groups),
        *_multi_package(buyable),
        *_near_values(buyable, percent),
        *_singletons(groups),
    ]


# `Description` and `Datasheet` are BOM columns like any other, and read best last.
_TRAILING_COLUMNS = ["Description", "Datasheet"]

# What separates two spellings' disagreeing values in a merged field:
# `RC0402FR-0710KL · RC0603FR-0710KL` is two parts arguing, not an MPN.
FIELD_JOIN = " · "


def _bom_fields(lines: list[dict]) -> dict[str, str]:
    """The BOM's own columns for one part, disagreements joined rather than hidden."""
    collected: dict[str, list[str]] = {}

    def add(name: str, value: str | None) -> None:
        trimmed = (value or "").strip()
        if trimmed == "":
            return
        seen = collected.get(name)
        if seen is None:
            collected[name] = [trimmed]
        elif trimmed not in seen:
            seen.append(trimmed)

    for line in lines:
        for name, value in line["line"]["fields"].items():
            add(name, value)
        add("Description", line["line"]["description"])
        add("Datasheet", line["line"]["datasheet"])

    return {name: FIELD_JOIN.join(collected[name]) for name in js_sorted(collected.keys(), compare_columns)}


def compare_columns(a: str, b: str) -> int:
    """Custom fields first, alphabetically and case-insensitively; the long prose last."""

    def rank(name: str) -> int:
        return _TRAILING_COLUMNS.index(name) + 1 if name in _TRAILING_COLUMNS else 0

    by_rank = rank(a) - rank(b)
    if by_rank:
        return by_rank
    by_name = locale_compare(a, b, base=True)
    if by_name:
        return by_name
    return -1 if a < b else 1 if a > b else 0


def parts_by_key(lines: Iterable[dict]) -> list[dict]:
    """One row per distinct part."""
    rows = []
    for group in _group_by_key(line for line in lines if line["excluded"] is None):
        first = group["lines"][0]
        spec = first["spec"]
        rows.append(
            {
                "key": group["key"],
                "label": search_label(spec),
                "cls": spec["cls"],
                "value": format_magnitude(spec["magnitude"]) if spec["kind"] == "generic" else spec["designator"],
                "pkg": spec["pkg"] if spec["pkg"] is not None else "?",
                "placements": group["placements"],
                "refs": group["refs"],
                "sources": [
                    {
                        "value": line["line"]["value"].strip(),
                        "footprint": line["line"]["footprint"],
                        "refs": line["line"]["refs"],
                        "dnp": line["line"]["dnp"],
                    }
                    for line in group["lines"]
                ],
                "issueCount": sum(len(line["issues"]) for line in group["lines"]),
                "fields": _bom_fields(group["lines"]),
                "resolved": spec["kind"] == "specific" and any(line["resolved"] for line in group["lines"]),
            }
        )
    # most placements first, then by key as localeCompare orders it
    return sorted(rows, key=lambda p: (-p["placements"], collation_key(p["key"])))


def distinct_parts(lines: Iterable[dict]) -> int:
    """Distinct parts once every free merge has been taken."""
    return len({line["key"] for line in lines})

File: src/core/consolidate/findings.ts
import { canonicalKey, searchLabel } from "../canonical.js";
import type { ComponentClass } from "../parse/refdes.js";
import type { AnalyzedLine, Spec } from "../types.js";
import { formatMagnitude, type Unit } from "../units.js";

export type PackageGroup = {
  pkg: string;
  placements: number;
  refs: string[];
};

export type Finding =
  /** Two BOM lines that spell the same part differently. Free to merge. */
  | {
      kind: "duplicate-spelling";
      key: string;
      spellings: { value: string; placements: number; refs: string[] }[];
      placements: number;
      linesSaved: number;
    }
  /** One part, one package, but reached through different footprints. */
  | {
      kind: "mixed-footprint";
      key: string;
      footprints: { footprint: string; placements: number; refs: string[] }[];
    }
  /** One value bought in several sizes. The user's original question. */
  | {
      kind: "multi-package";
      cls: ComponentClass;
      value: string;
      unit: Unit;
      groups: PackageGroup[];
      /** Package carrying the most placements — the cheapest standard to adopt. */
      suggested: string;
      /** True when the leading packages are equally common, so `suggested` is arbitrary. */
      tied: boolean;
      linesSaved: number;
    }
  /** Values so close together that having both is probably an accident. */
  | {
      kind: "near-value";
      cls: ComponentClass;
      members: { key: string; value: string; placements: number; refs: string[] }[];
      spreadPercent: number;
    }
  /**
   * A *generic* value used exactly once: a candidate for substitution.
   * Specific parts are excluded — a board with one MCU is not a finding.
   */
  | {
      kind: "singleton";
      key: string;
      cls: ComponentClass;
      value: string;
      ref: string;
    };

export type ConsolidationOptions = {
  /** Relative difference under which two values are "suspiciously close". */
  nearValuePercent?: number;
};

type Grouped = {
  key: string;
  lines: AnalyzedLine[];
  placements: number;
  refs: string[];
};

function groupByKey(lines: readonly AnalyzedLine[]): Grouped[] {
  const map = new Map<string, AnalyzedLine[]>();
  for (const line of lines) {
    const bucket = map.get(line.key);
    if (bucket) bucket.push(line);
    else map.set(line.key, [line]);
  }
  return [...map.entries()].map(([key, ls]) => ({
    key,
    lines: ls,
    placements: ls.reduce((n, l) => n + l.line.refs.length, 0),
    refs: ls.flatMap((l) => l.line.refs),
  }));
}

/** Generic specs only, keyed by class and magnitude so packages can be compared. */
function valueKey(spec: Extract<Spec, { kind: "generic" }>): string {
  return `${spec.cls}|${formatMagnitude(spec.magnitude)}`;
}

function duplicateSpellings(groups: readonly Grouped[]): Finding[] {
  const out: Finding[] = [];
  for (const group of groups) {
    const byValue = new Map<string, AnalyzedLine[]>();
    for (const line of group.lines) {
      const value = line.line.value.trim();
      const bucket = byValue.get(value);
      if (bucket) bucket.push(line);
      else byValue.set(value, [line]);
    }
    if (byValue.size < 2) continue;

    out.push({
      kind: "duplicate-spelling",
      key: group.key,
      spellings: [...byValue.entries()]
        .map(([value, ls]) => ({
          value,
          placements: ls.reduce((n, l) => n + l.line.refs.length, 0),
          refs: ls.flatMap((l) => l.line.refs),
        }))
        .sort((a, b) => b.placements - a.placements),
      placements: group.placements,
      linesSaved: byValue.size - 1,
    });
  }
  return out;
}

function mixedFootprints(groups: readonly Grouped[]): Finding[] {
  const out: Finding[] = [];
  for (const group of groups) {
    const byFootprint = new Map<string, AnalyzedLine[]>();
    for (const line of group.lines) {
      const fp = line.line.footprint.trim();
      const bucket = byFootprint.get(fp);
      if (bucket) bucket.push(line);
      else byFootprint.set(fp, [line]);
    }
    if (byFootprint.size < 2) continue;

    out.push({
      kind: "mixed-footprint",
      key: group.key,
      footprints: [...byFootprint.entries()]
        .map(([footprint, ls]) => ({
          footprint,
          placements: ls.reduce((n, l) => n + l.line.refs.length, 0),
          refs: ls.flatMap((l) => l.line.refs),
        }))
        .sort((a, b) => b.placements - a.placements),
    });
  }
  return out;
}

function multiPackage(lines: readonly AnalyzedLine[]): Finding[] {
  const byValue = new Map<string, AnalyzedLine[]>();
  for (const line of lines) {
    if (line.spec.kind !== "generic") continue;
    const k = valueKey(line.spec);
    const bucket = byValue.get(k);
    if (bucket) bucket.push(line);
    else byValue.set(k, [line]);
  }

  const out: Finding[] = [];
  for (const ls of byValue.values()) {
    const byPkg = new Map<string, AnalyzedLine[]>();
    for (const line of ls) {
      const pkg = line.spec.pkg ?? "?";
      const bucket = byPkg.get(pkg);
      if (bucket) bucket.push(line);
      else byPkg.set(pkg, [line]);
    }
    if (byPkg.size < 2) continue;

    const groups: PackageGroup[] = [...byPkg.entries()]
      .map(([pkg, group]) => ({
        pkg,
        placements: group.reduce((n, l) => n + l.line.refs.length, 0),
        refs: group.flatMap((l) => l.line.refs),
      }))
      .sort((a, b) => b.placements - a.placements || a.pkg.localeCompare(b.pkg));

    const first = ls[0] as AnalyzedLine;
    const spec = first.spec as Extract<Spec, { kind: "generic" }>;
    out.push({
      kind: "multi-package",
      cls: spec.cls,
      value: formatMagnitude(spec.magnitude),
      unit: spec.unit,
      groups,
      suggested: (groups[0] as PackageGroup).pkg,
      // A tie means the tool has no basis to prefer either size; say so rather
      // than presenting an arbitrary pick as a recommendation.
      tied: groups.length > 1 && groups[0]?.placements === groups[1]?.placements,
      linesSaved: groups.length - 1,
    });
  }
  return out;
}

type Magnitudes = Map<number, { refs: string[]; keys: Set<string> }>;

/**
 * Groups by class and then by *magnitude*, deliberately not by key: 10k in 0402
 * and 10k in 0603 are the multi-package finding, not a near-value one.
 */
function magnitudesByClass(lines: readonly AnalyzedLine[]): Map<ComponentClass, Magnitudes> {
  const byClass = new Map<ComponentClass, Magnitudes>();

  for (const line of lines) {
    // 0R has no meaningful neighbourhood: everything is infinitely far from it.
    if (line.spec.kind !== "generic" || line.spec.magnitude === 0) continue;

    let magnitudes = byClass.get(line.spec.cls);
    if (!magnitudes) {
      magnitudes = new Map();
      byClass.set(line.spec.cls, magnitudes);
    }

    let slot = magnitudes.get(line.spec.magnitude);
    if (!slot) {
      slot = { refs: [], keys: new Set<string>() };
      magnitudes.set(line.spec.magnitude, slot);
    }
    slot.refs.push(...line.line.refs);
    slot.keys.add(line.key);
  }

  return byClass;
}

function nearValues(lines: readonly AnalyzedLine[], percent: number): Finding[] {
  const out: Finding[] = [];

  for (const [cls, magnitudes] of magnitudesByClass(lines)) {
    const sorted = [...magnitudes.entries()].sort((a, b) => a[0] - b[0]);

    // Walk the sorted values, collecting runs where each step is within
    // tolerance of the last. 49.9k -> 50k -> 51k chains into one finding.
    let run: typeof sorted = [];
    const flush = (): void => {
      const lo = run[0]?.[0];
      const hi = run[run.length - 1]?.[0];
      if (run.length >= 2 && lo !== undefined && hi !== undefined) {
        out.push({
          kind: "near-value",
          cls,
          members: run.map(([magnitude, slot]) => ({
            key: [...slot.keys].sort().join(" "),
            value: formatMagnitude(magnitude),
            placements: slot.refs.length,
            refs: slot.refs,
          })),
          spreadPercent: ((hi - lo) / lo) * 100,
        });
      }
      run = [];
    };

    for (const entry of sorted) {
      const previous = run[run.length - 1];
      if (previous && ((entry[0] - previous[0]) / previous[0]) * 100 <= percent) {
        run.push(entry);
      } else {
        flush();
        run = [entry];
      }
    }
    flush();
  }

  return out;
}

function singletons(groups: readonly Grouped[]): Finding[] {
  const out: Finding[] = [];
  for (const group of groups) {
    if (group.placements !== 1) continue;
    const line = group.lines[0] as AnalyzedLine;
    // Only generics: you can swap a lone 82k for an 80.6k you already stock,
    // but a board with one RP2040 is not a consolidation opportunity.
    if (line.spec.kind !== "generic") continue;
    out.push({
      kind: "singleton",
      key: group.key,
      cls: line.spec.cls,
      value: line.line.value.trim(),
      ref: group.refs[0] as string,
    });
  }
  return out;
}

export function consolidate(
  lines: readonly AnalyzedLine[],
  options: ConsolidationOptions = {},
): Finding[] {
  const percent = options.nearValuePercent ?? 2;
  // Excluded lines never reach the report: a board with 8 test points should
  // not produce eight "used once" findings about pads you cannot buy.
  const buyable = lines.filter((l) => l.excluded === undefined);
  const groups = groupByKey(buyable);

  return [
    ...duplicateSpellings(groups),
    ...mixedFootprints(groups),
    ...multiPackage(buyable),
    ...nearValues(buyable, percent),
    ...singletons(groups),
  ];
}

export type PartRollup = {
  key: string;
  /**
   * The same part as a distributor writes it — `CAP CER 0.1UF 0402` — for
   * pasting into a search box. A label, never an identity: `key` is what
   * anything is filed under.
   */
  label: string;
  cls: ComponentClass;
  value: string;
  pkg: string;
  placements: number;
  refs: string[];
  /** The BOM lines that fed this part, so differing spellings stay visible. */
  sources: { value: string; footprint: string; refs: string[]; dnp: boolean }[];
  issueCount: number;
  /** Already a specific part: nothing for step 2 to resolve. */
  resolved: boolean;
  /**
   * Everything else the BOM said about this part: `Description`, `Datasheet`
   * and every custom field the designer maintains, merged over the lines that
   * fed the key. Two spellings that disagree keep both values rather than one
   * of them winning silently.
   */
  fields: Record<string, string>;
};

/** `Description` and `Datasheet` are BOM columns like any other, and read best last. */
const trailingColumns = ["Description", "Datasheet"];

/**
 * What separates two spellings' disagreeing values in a merged field.
 *
 * Exported because a reader has to be able to tell a value from a question:
 * `RC0402FR-0710KL · RC0603FR-0710KL` is two parts arguing, not an MPN.
 */
export const fieldJoin = " · ";

/**
 * The BOM's own columns for one part.
 *
 * Distinct values are joined rather than reduced to the first: when `100n` and
 * `100nF` merge into one part and only one of them carries `Volrtage 50V`, the
 * disagreement is exactly what the row should show.
 */
function bomFields(lines: readonly AnalyzedLine[]): Record<string, string> {
  const collected = new Map<string, string[]>();
  const add = (name: string, value: string | undefined): void => {
    const trimmed = (value ?? "").trim();
    if (trimmed === "") return;
    const seen = collected.get(name);
    if (seen === undefined) collected.set(name, [trimmed]);
    else if (!seen.includes(trimmed)) seen.push(trimmed);
  };

  for (const l of lines) {
    for (const [name, value] of Object.entries(l.line.fields)) add(name, value);
    add("Description", l.line.description);
    add("Datasheet", l.line.datasheet);
  }

  const fields: Record<string, string> = {};
  for (const name of [...collected.keys()].sort(compareColumns)) {
    fields[name] = (collected.get(name) as string[]).join(fieldJoin);
  }
  return fields;
}

/** Custom fields first, alphabetically and case-insensitively; the long prose last. */
export function compareColumns(a: string, b: string): number {
  const rank = (name: string): number => trailingColumns.indexOf(name) + 1;
  return (
    rank(a) - rank(b) ||
    a.localeCompare(b, "en", { sensitivity: "base" }) ||
    (a < b ? -1 : a > b ? 1 : 0)
  );
}

/** One row per distinct part — what the inventory will eventually be built from. */
export function partsByKey(lines: readonly AnalyzedLine[]): PartRollup[] {
  return groupByKey(lines.filter((l) => l.excluded === undefined))
    .map((group) => {
      const first = group.lines[0] as AnalyzedLine;
      return {
        key: group.key,
        label: searchLabel(first.spec),
        cls: first.spec.cls,
        value:
          first.spec.kind === "generic"
            ? formatMagnitude(first.spec.magnitude)
            : first.spec.designator,
        pkg: first.spec.pkg ?? "?",
        placements: group.placements,
        refs: group.refs,
        sources: group.lines.map((l) => ({
          value: l.line.value.trim(),
          footprint: l.line.footprint,
          refs: l.line.refs,
          dnp: l.line.dnp,
        })),
        issueCount: group.lines.reduce((n, l) => n + l.issues.length, 0),
        fields: bomFields(group.lines),
        resolved: first.spec.kind === "specific" && group.lines.some((l) => l.resolved),
      };
    })
    .sort((a, b) => b.placements - a.placements || a.key.localeCompare(b.key));
}

/** Distinct parts once every free merge has been taken. */
export function distinctParts(lines: readonly AnalyzedLine[]): number {
  return new Set(lines.map((l) => l.key)).size;
}

export { canonicalKey };
